src/openfermion/ops/_symbolic_operator.py

```diff
@@ -96,13 +96,64 @@ def __init__(self, term=None, coefficient=1.):
         # Add the term to the dictionary
         self.terms[term] = coefficient
 
-    def _long_string_init(self, term, coefficient):
+    def _long_string_init(self, long_string, coefficient):
         """
-        Initialization from a long string representation, i.e., term
-        is a string such as '1.5 [2^ 3] + 2.4 [3^ 0]'.
+        Initialization from a long string representation.
+
+        e.g. For FermionOperator:
+            '1.5 [2^ 3] + 1.4 [3^ 0]'
         """
-        raise NotImplementedError('Initializing a SymbolicOperator from a '
-                                  'long string is not yet supported.')
+
+        index = 0  # Index of current character in long_string
+        subtraction = False  # Whether this term is being subtracted
+        while index < len(long_string):
```

Comment:
Is there a good reason to use while loops that increment an index as opposed to using for loops and "enumerate"? Unless there is a good reason, I think for loops and enumerate tend to lead to code that is bit easier to parse.

Reply:
I've changed it now to loop over regex matches.